APE21.rst

```diff
@@ -0,0 +1,105 @@
+Ending Long Term Support Releases
+---------------------------------
+
+author: Thomas Robitaille, Pey Lian Lim, Simon Conseil
+
+date-created: 2023 February 16
+
+date-last-revised: 2023 February 21 <keep this up to date anytime something changes>
+
+date-accepted: <replace with accepted date>
+
+type: Process
+
+status: Discussion
+
+
+Abstract
+--------
+
+This APE proposes to no longer designate any releases of the core package as being
+"long-term support" (LTS) and provide guidance on how to handle API changes and deprecations.
+
+
+Detailed description
+--------------------
+
+`APE 2: Astropy Release Cycle and Version Numbering <https://doi.org/10.5281/zenodo.1043887>`_ describes the release cycle
+for the core astropy package, which consists of major releases every 6 months that
+add new features (e.g., v5.2), and releases in between this that are limited to fixing bugs
+or updating documentation (e.g., v5.2.1). Every 2 years, major releases are designated
+as being LTS (e.g., v5.0) and we have committed to releasing bugfix releases
+for this version (e.g., v5.0.3) for up to 2 years.
+
+At the start of the project, having LTS releases was critical for acceptance by
+large organizations/projects for which stability is paramount, as the core package went through
+significant API changes from version to version due to the youth of the project and LTS
+releases were a way to shield against these changes, or at least reduce disruption to
+be every 2 years.
+
+LTS releases are however not free:
+
+* While we use infrastructure to automate backporting fixes to release branches,
+  in practice the LTS branch diverges more and more from the main development
+  branch and automated backports are more likely to fail as time goes on,
+  requiring human effort to manually do the backports, which can sometimes be
+  non-trivial.
+* Even once fixes are backported to the LTS branch, carrying out releases is not
+  fully automated and requires manual steps, so having to do
+  bugfix releases for both the latest release branch and LTS branch puts an extra
+  burden on the core package release team.
+
+In addition, it is not clear that LTS releases are needed anymore - now that the
+core package is over ten years old, it has reached a much higher level of
+stability. No other main package in the scientific Python ecosystem has LTS
+releases, and there is no longer a strong argument for having this
+in astropy either.
+
+This APE proposes the following:
+
+* No longer designate any releases as LTS.
+* Keep the 6-month release cycle and bump releases to x.0 every 2 years.
```

Comment:
But this also means that if someone wants a new thing that will totally break API, they have to wait 2 years? Do we not merge that PR for 2 years?

Reply:
Yeah indeed. This case should hopefully not happen too often, but we could imagine releasing a major version only 1/1.5 year after the previous one on exceptional circumstances ? But in general if there is a need to totally break the API there should be a long enough period of time before that with deprecation warnings, so the 2 years cycle seem realistic.

Reply:
It's worth noting that with a fixed 2-year cycle, some API changes might wait 1.9 years, and others might be merged 1 week before the deadline.

There would tend to be a bias (maybe strong?) toward the major version feature freeze deadline, because folks might have an idea and want to get it in knowing that it will be another 2 years otherwise. Something to consider!

Reply:
Yes, that's why I think we should keep some flexibility in the 2 years cycle. 
But also before a breaking change appears in a major release, there should be deprecations in a previous minor release. So major release are not the time to merge anything just because we can break things in this release, we should plan beforehand what major changes are waiting, if deprecations have been released, for how long, etc.

Reply:
Agree with @taldcroft here. The most annoying numpy changes we have had to deal with are things that are merged just before the release and hence we didn't catch them in `-dev`.

Would it be an idea to freeze further ahead of a major release? With perhaps more advertising to test the corresponding -rc?

Alternatively (or additionally), maybe the way to do this is to be more flexible about when major releases happen, and decide on whether the next release will be major ~mid-way between releases, based on what is already ready there and what is being planned (with perhaps an astropy-dev mailing to ask what people are planning).

Part of this might also be to follow pandas' example and really make sure API changes are explained and work-arounds given.

One practical question: how is it going to work with major items? Are they just going to sit on as PR requests for ~1 year? Or is there a separate branch? As long as new features are OK for minor releases, I don't think this is a super-big issue, but probably should be discussed in the document!

Reply:
Should have read @saimn's message - agreed that if the cycle is flexible and we insist (as we should) that breaking changes have deprecation periods, that solves most of the problems.

Reply:
> One practical question: how is it going to work with major items? Are they just going to sit on as PR requests for ~1 year? Or is there a separate branch? As long as new features are OK for minor releases, I don't think this is a super-big issue, but probably should be discussed in the document!

Yes new features are OK for minor releases. It's just the ones that break the API that should wait, with deprecation added in between. And that should be a minority of the PRs hopefully, so I guess we can just milestone them for the next major release and rebase when we start the major release cycle. Another way would be to have a "future" branch but if there are conflicts it will be more painful to resolve them compared to doing that for each PR.